### PR TITLE
[DO NOT MERGE] Exploring slight changes to APIs

### PR DESCRIPTION
**Pull request by ktoso**
Not intended for merging but showcasing rationale for "decide type for metadata value #5" as well as  some minor refactor proposals for the api:

- potentially change `public var metadata: LoggingMetadata? {` to `public var metadata: LoggingMetadata {` to avoid nil compensation code everywhere
- avoid "eager rendering" of values which need to compute their string representations by:
  - avoid `String` value type in the Metadata dictionary
  - avoid `didSet` and storing (pre)rendered prettyMetadata since it forces eagerly rendering the values, even if logging does not happen with current metadata
  - only render metadata when indeed logging values

This PR is not intended to be merged as is, but only to show the main idea behind a change.
If we agree on the changes I can make a clean PR or if you'd rather apply changes we'd agree on yourself I'm find with that happening independently as well.

To run the example and observe measurements (not super advanced, but hey, it's something):

`swift run Examples --configuration release`

### CODE REVIEW

File: Sources/Examples/ExploringPerformanceExample.swift [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+import Logging
+
+// MARK: Example library code
+
+struct DataContainer {
+    let data: String // this would be some user `T` type
+
+    // would be W3C's `traceparent` for example
+    let traceContext: TraceContext?
+}
```
Comment (ktoso) [inline Sources/Examples/ExploringPerformanceExample.swift:11]: This is just an example, though showcases a series of cases where "we want to carry around a nice binary representation and only render if when we really need to", in this case it could be a trace context from which we'd for example only want to log the `traceId` for correlation purposes of spans with logs.

(yes, tracing is a separate domain entirely, however correlating with logging is very useful, and since "tons of things" would have this, it would be good to not pay penalties for it when we don't actually log with them).

File: Sources/Examples/ExploringPerformanceExample.swift [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    public var description: String {
+        let v = String(self.version, radix: 16, uppercase: true)
+        let t1 = String(self.traceId.0, radix: 16, uppercase: true)
+        let t2 = String(self.traceId.1, radix: 16, uppercase: true)
+        let s = String(self.parentId, radix: 16, uppercase: true)
+        let f = String(self.flags, radix: 16, uppercase: true)
+
+        // rendering just an example, did not super closely follow any spec, but close enough:
+        return "\(v)-\(t1)\(t2)-\(s)-\(f)"
+    }
```
Comment (ktoso) [inline Sources/Examples/ExploringPerformanceExample.swift:35]: "expensive rendering" example; It's not super super expensive, but if you imagine there being many values which have an efficient representation which then need to print themselfes constantly even if we don't actually do print them them you get a feel of how big the waste is.

File: Sources/Examples/ExploringPerformanceExample.swift [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+    private func extractMeta(from envelope: DataContainer, withBase metadata: LoggingMetadata) -> LoggingMetadata {
+        var renderedMetadata: LoggingMetadata = [:]
+
+        switch envelope.traceContext {
+        case .some(let trace):
+            renderedMetadata = [
+                "per-data0": "some-metadata",
+                "trace": trace // by allowing Any as value, we delay the rendering operation
```
Comment (ktoso) [inline Sources/Examples/ExploringPerformanceExample.swift:74]: this rides on the fact that LoggingMetadata is now `String:Any` so we don't have to invoke the rendering of the trace yet (nor any other values which would have to run some more code to render than a plain "expose my fields").

With a `String:String` metadata container we'd be forced to:

```swift
        var renderedMetadata: LoggingMetadata = [:]

         switch envelope.traceContext {
         case .some(let trace):
             renderedMetadata = [
                 "per-data0": "some-metadata",
                 // TODO: So that's the thing I'm a bit worried about with the string metadata:
                 // TODO: this rendering is a bit costly, though we'd likely want to carry around the integers rather than the string repr...
                 // TODO: Sadly with String metadata we always have to render it, even if we don't even log here AT ALL :-(
                 // TODO: solution 1) pass around the trace metadata as String all the time in the envelope, this
                 // TODO: solution 2) "whatever", yeah it costs
                 "trace": "\(trace)" // runs . description eagerly, even if we'd never log
             ]
```

(such impl is in 3ae6ec6718e04308c2b48bd9dc4ad36a45645ced )

File: Sources/Examples/ExploringPerformanceExample.swift [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    func handle(userCallback: (inout SomeContext, DataContainer) -> ()) {
+        let perContextMetadata: LoggingMetadata = context.log.metadata
+
+        let trace = TraceContext( // assume we have it, it came with the container
+            version: 0,
+            traceId: (790211418057950173, 9532127138774266268),
+            parentId: 67667974448284343,
+            flags: 0
+        )
+        let data = DataContainer(data: "example", traceContext: trace)
```
Comment (ktoso) [inline Sources/Examples/ExploringPerformanceExample.swift:95]: // data would come in from the network or somewhere else, here I just mock it for example purposes

File: Sources/Logging/Logging.swift [hunk 5]
```diff
@@ -13,7 +13,7 @@ public protocol LogHandler {
     subscript(metadataKey _: LoggingMetadata.Key) -> LoggingMetadata.Value? { get set }
 
     // All available metatdata
-    var metadata: LoggingMetadata? { get set }
+    var metadata: LoggingMetadata { get set } // removal of `?`, don't see the gain vs. using an empty map
```
Comment (ktoso) [inline Sources/Logging/Logging.swift:16]: I see that `LoggingMetadata?` enabled performing a `set` with `nil` in order to work like a "clear" though I'm not entirely sure this is so nice, and we end up sprinkling a lot of `?` and `!` in logger handlers due to this.

How about we define a "clear context" as `= LoggingMetadata()` which is `= [:]` which should be fine and avoids all those `nil` compensating codepaths.

File: Sources/Logging/Logging.swift [hunk 6]
```diff
@@ -62,7 +62,7 @@ public struct Logger {
     }
 
     @inlinable
-    public subscript(metadataKey metadataKey: String) -> String? {
+    public subscript(metadataKey metadataKey: LoggingMetadata.Key) -> LoggingMetadata.Value? {
```
Comment (ktoso) [inline Sources/Logging/Logging.swift:65]: nitpick, reads a bit nicer

File: Sources/Logging/Logging.swift [hunk 7]
```diff
@@ -100,7 +100,28 @@ public enum LogLevel: Int {
...
+//    10ms 915μs, 6ms 531μs, 6ms 68μs, 5ms 735μs, 5ms 870μs, 9ms 493μs, 6ms 564μs, 7ms 400μs, 10ms 455μs, 7ms 205μs,
+//    measuring: no 10000 logs:
+//    1ms 643μs, 1ms 592μs, 1ms 671μs, 1ms 579μs, 1ms 602μs, 1ms 565μs, 1ms 560μs, 1ms 539μs, 1ms 510μs, 1ms 499μs,
+//
+// String:Any, avoiding didSet, rendering only when actually printing
+//    measuring: 100 logs:
+//    // [2018-12-17 12:04:45 +0000][info][additional=some-per-context-info trace=0-AF7651916CD43DD8448EB211C80319C-F067AA0BA902B7-0 per-data0=some-metadata id=some-id] Hello 100
+//    20ms 89μs, 6ms 877μs, 8ms 257μs, 10ms 628μs, 6ms 299μs, 5ms 718μs, 5ms 723μs, 5ms 362μs, 8ms 759μs, 6ms 155μs,
+//    measuring: no 10000 logs:
+//    675μs 82ns, 394μs 940ns, 393μs 986ns, 393μs 986ns, 393μs 986ns, 394μs 940ns, 395μs 59ns, 394μs 940ns, 395μs 59ns, 486μs 16ns,
```
Comment (ktoso) [inline Sources/Logging/Logging.swift:130]: These are the results I'm trying to make nice; cases where we propagate those "expensive to render" things, but we never actually print so it does not need to be rendered into its string representation.

Note that when we DO print them, we may indeed print them a few times... but we could work around this with some caching of the "heavy reprs"? I'm more worried about the trace/debug logs not impacting performance when they're off, but causing rendering of the strings anyway, since the metadata object forcing them to, even if loglevel would mean they should not ever appear.

File: Tests/LoggingTests/TestLogger.swift [hunk 8]
```diff
@@ -31,7 +31,10 @@ internal struct TestLoggger: LogHandler {
     }
 
     func log(level: LogLevel, message: String, file: String, function: String, line: UInt) {
-        let metadata = self.metadata ?? MDC.global.metadata // use MDC unless set
+        // I don't think this would be a correct way of using those, rather a merge is needed:
+        // let metadata = self.metadata ?? MDC.global.metadata // use MDC unless set
+        let metadata = MDC.global.metadata.merging(self.metadata, uniquingKeysWith: { l, r in r })
```
Comment (tomerd) [inline Tests/LoggingTests/TestLogger.swift:36]: @ktoso being able to choose between the local metadata (when set, empty or not) or falling back to some globally defined one (like the MDC in this example) is one of the key reasons for choosing metadata to be optional. here you solved it by a merge, but my design assumption was that merge is not always what you want. sometimes you'd want to be able and completely override the metadata. obviously we leave this to the LogHandler implementation, and the user can always do the merge at their level too. higher order bit is that without some metadata "have-been-set" state this will be impossible to do. that said, having metadata optional is one way to implement a "have-been-set" state, we can also add a didSet observable on metadata to achieve the same

File: Sources/Logging/Logging.swift [hunk 9]
```diff
@@ -144,20 +179,15 @@ public final class StdoutLogger: LogHandler {
...
-        didSet {
-            self.prettyMetadata = !(self._metadata?.isEmpty ?? true) ? self._metadata!.map { "\($0)=\($1)" }.joined(separator: " ") : nil
-        }
-    }
-
+    private var _metadata: LoggingMetadata
     public func log(level: LogLevel, message: String, file _: String, function _: String, line _: UInt) {
         if level >= self.logLevel {
-            print("\(Date()) \(level)\(self.prettyMetadata.map { " \($0)" } ?? "") \(message)")
+            let prettyMetadata = !(self._metadata.isEmpty) ? self._metadata.map { "\($0)=\($1)" }.joined(separator: " ") : ""
```
Comment (tomerd) [inline Sources/Logging/Logging.swift:185]: the idea of prettyMetadata above was that its only computed when metadata changes, instead of on the hot path

File: Sources/Logging/Logging.swift (hunk 6 above)
Comment (tomerd) [reply Sources/Logging/Logging.swift:65]: +1 i think we had it this way at some point maybe @weissi decided against it?

File: Sources/Logging/Logging.swift (hunk 5 above)
Comment (tomerd) [reply Sources/Logging/Logging.swift:16]: see comment below

File: Tests/LoggingTests/TestLogger.swift (hunk 8 above)
Comment (ktoso) [reply Tests/LoggingTests/TestLogger.swift:36]: Whoops, just opened https://github.com/weissi/swift-server-logging-api-proposal/issues/7 but will consider your idea here in depth and reply in a moment :) Thanks for chiming in :)

File: Sources/Logging/Logging.swift (hunk 9 above)
Comment (ktoso) [reply Sources/Logging/Logging.swift:185]: (Right also to the "hot path", see below though) Right, I agree that "rendering many times" is wasteful, however so is "rendering when we don't ever print" the pretty metadata, yet the field is changed a lot.

So the downsides of the 2 styles are:

- `prettyMetadata` String property, re-rendered at `metadata didSet` – we render all the values even if they end up not used. This is specifically hurting in the case of: used log level determines we would not log things, we set metadata per each execution, it is different because depends on the data (e.g. a trace's parent id), we render the prettyMetadata even though we don't know if we'll print it. 
  - a) we don't know because maybe there is no `log` call at all 
  - b) we don't know* because based on loglevel just now, maybe it should be printed... but then loglevel is changed, and maybe we now would print or now maybe we would not. So we'd have to inspect log level here, but also potentially trigger rendering when loglevel is changed. but even then we may render needlessly, as loglevel was changed, but nothing was logged – so we did a wasteful render of the `prettyMetadata` again.
- the "in this PR" proposal, renders only when we log – however it renders many many times, for each log call. This is also wasteful, if the logging never changed.

Note on hot path: I may be biased by my execution styles I work with, but the setting of the metadata is immediately right before the code executing user code, so it does not change the entire "time to process". It would change the "time spent in user handler" though, if we render before yes... I'm not convienced this matters in the end, as users look at "what's my throughput", and both executions happen in the same hot loop in either case.

I think we can achieve best of both worlds, perhaps. How about:

- keep the `prettyMetadata: String?` property
- in `metadata didSet { prettyMetadata = nil }`, so we invalidated our cached `prettyMetadata` since the values changed.
- user code runs:
  - case a) no logging happens, we never render this specific set of metadata (good)
  - case b) logging happens, on log level that does not trigger actual printing though – no rendering happens (good)
  - case c) logging happens, on log level that causes printing; if the `prettyMetadata` is `nil` we render once and initialize it; subsequent logging reuses the once rendered metadata – we avoided double rendering, and avoided rendering too early (good, in my book at least)
- some cases to consider when loglevel is changed in user code:
  - case a) log event happens, though it is `trace` and we only log from `info`, meaning we did not render the metadata, user code sets the log level to `trace` so we should now render it; the next log on `trace` causes the render to happen and caches it (good)
  - case b) user logs at various log levels, if at least one of them has to be printed we render; we don't render twice though (good)

So we get the cached `prettyMetadata` and also avoid rendering it in case no log event would make use of it. It also retains the benefit the initial `prettyMetadata` had where if metadata does not change we don't render it again.

WDYT?

Comment (ktoso) [reply Sources/Logging/Logging.swift:185]: (Note to self: so technically this is an implementation thing, however at the same time it relates to if/how we restrict the type of values to be Strings in the Metadata or not; if not, then these optimizations come into play; if they are forced to be strings already it matters less, as we were forced to make the rendering before already...)

// I'm looking at context APIs now and if/how they could allow avoiding these issues, though don't have an answer yet.

File: Tests/LoggingTests/TestLogger.swift (hunk 8 above)
Comment (ktoso) [reply Tests/LoggingTests/TestLogger.swift:36]: Thanks for clarifying Tom, it indeed seems to be about the different design assumptions we both operated on.

> here you solved it by a merge, but my design assumption was that merge is not always what you want. sometimes you'd want to be able and completely override the metadata. obviously we leave this to the LogHandler implementation, and the user can always do the merge at their level too.

Yeah, the merge with a global MDC perhaps is a not so fortunate application of this. I can agree that perhaps the local one may want to sometimes "win" etc... "it depends"™ is the killer key phrase here I guess. I agree that since this is only about the API really, we should keep it open enough to satisfy the use case as you lined it out as well, in case someone wants it.

For painting a fuller picture I want to explain a bit where the merging in my examples comes from: In applications I deal with the "hierarchy" is "natural" and always there, so lower levels are "more specific" one could say thus they may want to override, though I can see that if an application has no such layers perhaps it would care less.

Last example why the layers make most sense for me and some kinds of apps, layers being:
- "the context where you execute" which remains relatively stable, it contains some metadata which we may want to log,
- then per execution we have some "request dependent" metadata, the most typical examples would be anything that's a trace etc. this is still not really set by user code.
- user code may also set some metadata, on a logger we provided them, we do not want to remove these metadata, the users are in charge of clearing it.

So in this scheme "overriding" and "reset to the stable one" are the semantics I'm after.
The `Optional` does not block implementing this, but looked a bit weird to me, so if you have use cases which like keeping that optional information I'm happy to keep it in the API :-)

--- 

As we work on the general API here, I suppose keeping it open enough that both of our designs could work would be best, which in this case means keeping the `?` I suppose. I can live with that 😉 

TL;DR; Seems we keep the `?` to satisfy differentiating between "empty" and "did not set".
If an impl does not do not setting, it can always do [:] after all then.

**Comment by ktoso** [comment]
Closing since things got addressed in real PRs :)
